[stdlib] Make unsafe array initializer public

natecook1000 opened the pull request:
This is an implementation for the [re-pitch of SE-0223](https://github.com/apple/swift-evolution/pull/993).

The original underscored version of this initializer remains, with a deprecation notice. Also adds the initializer to `ContiguousArray`. Still to do:

- [x] Check inlining attributes, other annotations
- [x] Add tests for `ContiguousArray` and the non-underscored `Array` version


Resolves #45677.

natecook1000 committed:
[stdlib] Make unsafe array initializer public

The original underscored version of this initializer remains, with a
deprecation notice. Also add the initializer to ContiguousArray.

natecook1000 committed:
Update uninitialized array tests.

natecook1000 commented:
@swift-ci Please smoke test

natecook1000 commented:
@swift-ci Please smoke test

natecook1000 commented:
@swift-ci Please smoke test

natecook1000 commented:
@airspeedswift I was expecting this to fail the stable ABI test, since it’s adding new API — does that only track changes and deletions?

natecook1000 commented:
@swift-ci Please smoke test Linux platform

natecook1000 commented:
@swift-ci Please smoke test Linux platform

jrose-apple commented on stdlib/public/core/Array.swift:
```diff
@@ -1366,21 +1366,35 @@ extension Array {
 }
 
 extension Array {
+  @available(swift, deprecated: 5.1, renamed: "init(unsafeUninitializedCapacity:initializingWith:)")
```
This needs to be OS availability, unfortunately. It's not tied to `-swift-version` but to your deployment target.

natecook1000 replied:
Addressed these in b962048.

jrose-apple commented on stdlib/public/core/Array.swift:
```diff
@@ -1392,29 +1406,29 @@ extension Array {
   ///         elements you initialize.
   @inlinable
   public init(
```
…And this one *needs* OS availability.

tkremenek commented:
The corresponding evolution proposal has been accepted.

tkremenek commented:
@swift-ci clean test

natecook1000 committed:
Switch uninitialized Array inits to call original implementation

This strategy lets us avoid availability constraints on the new public
methods, since the original underscored version is part of the permanent
ABI.

natecook1000 committed:
Add additional tests for uninitialized Array init.

natecook1000 commented:
@swift-ci Please test

jrose-apple commented on stdlib/public/core/ContiguousArray.swift:
```diff
@@ -985,6 +985,43 @@ extension ContiguousArray {
+  ///       - initializedCount: The count of initialized elements in the array,
+  ///         which begins as zero. Set `initializedCount` to the number of
+  ///         elements you initialize.
+  @inlinable
```
Should this one also be `@_alwaysEmitIntoClient`?

natecook1000 replied:
Yes — fixed, thanks!

natecook1000 committed:
Add @_alwaysEmitIntoClient to ContiguousArray init.

natecook1000 commented:
@swift-ci Please test

Catfish-Man reviewed (approved):
This looks good to me!

natecook1000 committed:
Update CHANGELOG for Array unsafe initializer.

natecook1000 commented:
@swift-ci Please smoke test
